### ISSUE TITLE
Sphinx 1.4.8: LaTeX writer produce wrong notice block?

### ISSUE THREAD

**Issue by rexut** [sphinx-doc/sphinx#3117]
File [texinputs/sphinx.sty](https://github.com/sphinx-doc/sphinx/blob/1.4.8/sphinx/texinputs/sphinx.sty#L640) provides the LaTeX environment 'notice' that expect exactly 2 parameters. But the generator in file [writers/latex.py](https://github.com/sphinx-doc/sphinx/blob/1.4.8/sphinx/writers/latex.py#L1516) omit the second parameter completely when there is now special admonition label (ex. 'error', 'attention' or ...). Combined with a literal title, latex will run into a build error like this:

> ! Package amsmath Error: \check allowed only in math mode.
>
>See the amsmath package documentation for explanation.
>Type  H <return>  for immediate help.
> ...
>
>l.271 \begin{notice}{note}\sphinxcode
>                                     {ID-0815} -- Example Content\end{notice}

The LaTeX code is:
`\begin{notice}{note}\sphinxcode{ID-0815} -- Example Content\end{notice}`

... but should be (note the last empty `{}` for the begin statement):
`\begin{notice}{note}{}\sphinxcode{ID-0815} -- Example Content\end{notice}`

This behaviour was observed while using the [sphinxcontrib-traceables](https://github.com/t4ngo/sphinxcontrib-traceables) extension.

**Comment by jfbu** [comment]
Thanks for the report! 

But the admonition directive requires a title.

```
/xxx/sphinx/test-temptest/index.rst:11: ERROR: Error in "admonition" directive:
1 argument(s) required, 0 supplied.

.. admonition::

   ``ID-0815``
```

and if I try

```
.. admonition:: ``ID-0815``

   world
```

with a literal title I get (master branch)

``` latex
\begin{sphinxadmonition}{note}{\sphinxstyleliteralintitle{ID-0815}}

world
\end{sphinxadmonition}
```

or (1.4.8):

``` latex
\begin{notice}{note}{\texttt{ID-0815}}

world
\end{notice}
```

as latex code which has no issue. For info, not much has changed here since Sphinx 1.0 (the sphinx.sty handling of notice environment has changed recently but number of parameters has always been 2, the latex writer has almost not changed).

**Comment by rexut** [comment]
Yes, the admonition directive will block without a title. But I was running into this issue through the sphinxcontrib-traceables extension and it doesn't generate rst syntax. As far as I can see it right in the extension code base, it renders the admonition block for it self:
- [TraceableDisplayAdmonitionFormatter::format()](https://github.com/t4ngo/sphinxcontrib-traceables/blob/master/sphinxcontrib/traceables/display.py#L99)
- [TraceableDisplayFormatterBase::create_title_node()](https://github.com/t4ngo/sphinxcontrib-traceables/blob/master/sphinxcontrib/traceables/display.py#L70)

What ever, back to the Sphinx code. For my opinion there is a potential bug inside the LaTeX writer. The style defines the new 'notice' environment that expect 2 arguments in its interface. But the writer consists of a code fragment that is able to generate LaTeX syntax with one argument -- that is a bug, or not?

the style defines:

```
\newenvironment{notice}[2]{
  \def\py@noticetype{#1}
  \csname py@noticestart@#1\endcsname
  \sphinxstrong{#2} % <- legacy code creates a space after {#2}
}{\csname py@noticeend@\py@noticetype\endcsname}
```

in contrast to the writer:

```
    def visit_admonition(self, node):
        self.body.append('\n\\begin{notice}{note}')
```

The correct body append have to be `self.body.append('\n\\begin{notice}{note}{}')` in case of a missing title (for whatever reason the title is missing, that is not the problem here).

I would like to present a PR to fix this issue back to all historical releases that have this issue (bug?) inside. But this assumes that everyone agrees that we talk about a bug.

**Comment by jfbu** [comment]
I don't think there is a bug. The admonition directive requires a title argument. The title ends up braced in the tex document. The notice environment is always equipped with its two mandatory arguments. The situation has been like this since Sphinx 1.0.

```
hello

.. admonition:: x

   ``ID-0815``

   world
```

gives

``` latex
hello

\begin{sphinxadmonition}{note}{x}

\sphinxcode{ID-0815}

world
\end{sphinxadmonition}
```

incidentally: as you see 1.5 will use "sphinxadmonition". But for backwards compatibility this is just a wrapper of "notice":

```
\newenvironment{sphinxadmonition}{\begin{notice}}{\end{notice}}
```

**Comment by jfbu** [comment]
Consider this input:

```
hello

.. admonition:: \

   ``ID-0815``

   world
```

using `pseudoxml` output we get

``` xml
        <paragraph>
            hello
        <admonition classes="admonition-">
            <title>
            <paragraph>
                <literal>
                    ID-0815
            <paragraph>
                world
```

confirming the empty title and the latex writer produces:

``` latex
hello

\begin{sphinxadmonition}{note}{}

\sphinxcode{ID-0815}

world
\end{sphinxadmonition}
```

The empty title was correctly passed with a `{}` to the sphinxadmonition (aka notice) environment. The issue should be raised at the https://github.com/t4ngo/sphinxcontrib-traceables site because imho it is not a Sphinx issue.

**Comment by rexut** [comment]
Again, on RST level I'm completely with you, but not on level of the LaTeX writer. Almost no one has had seen this issue because of the admonition directive requires a title argument. But since Sphinx 1.0 there is a code fragment inside the LaTeX writer that will produce wrong LaTeX syntax:

Only in rarely cases this (wrong) code will execute:

```
    def visit_admonition(self, node):
        self.body.append('\n\\begin{notice}{note}')
```

In most of all cases (proper admonition directive) this code will execute:

```
    def _make_visit_admonition(name):
        def visit_admonition(self, node):
            self.body.append(u'\n\\begin{notice}{%s}{%s:}' %
                             (name, admonitionlabels[name]))
        return visit_admonition
```

**Comment by jfbu** [comment]
No the `visit_admonition` code will get executed each time a user correctly uses the generic adminition directive.

**Comment by rexut** [comment]
Why it will be defined separately inside of _make_visit_admonition() ?

**Comment by jfbu** [comment]
replace in latex.py "note" by "ZOUZOU" as below

```
def visit_admonition(self, node):
    self.body.append('\n\\begin{sphinxadmonition}{ZOUZOU}')
```

the do make latex on rst input like

``` rst
hello

.. admonition:: mytitle

   ``ID-0815``

   world
```

and you will see that the `visit_admonition` that you say is buggy is executed and does note create problematic output.

**Comment by jfbu** [comment]
by the way when I say it is like this since Sphinx 1.0 I do not mean to induce the feature was introduced then. It just means I did git checkout 1.0 and looked into latex.py and sphinx.sty.

**Comment by rexut** [comment]
Now I'm completely confused. I'm using Sphinx (1.4.8) with following test rst:

```
.. http://docutils.sourceforge.net/docs/ref/rst/directives.html#generic-admonition

Generic Admonition example from Docutils

.. admonition:: And, by the way...

   You can make up your own admonition too.

A little bit changed to use a literal string inside

.. admonition:: And, by the way...

   You can make up your own admonition too.
```

Then I call `sphinx-build -W -v -b latex -d build/doctrees . build/latex && cat build/latex/*.tex` and get this:

```
Generic Admonition example from Docutils

\begin{notice}{note}
You can make up your own admonition too.
\end{notice}

A little bit changed to use a literal string inside

\begin{notice}{note}{And, by the way...}

You can make up your own admonition too.
\end{notice}
```

The second one is right but the first one is buggy, why this?

After this I've replaced `visit_admonition()` in `writers/latex.py` to this:

```
    def visit_admonition(self, node):
        self.body.append('\n\\begin{notice}{ZOUZOU}')
```

... and run Sphinx again:

```
Generic Admonition example from Docutils

\begin{notice}{ZOUZOU}
You can make up your own admonition too.
\end{notice}

A little bit changed to use a literal string inside

\begin{notice}{ZOUZOU}{And, by the way...}

You can make up your own admonition too.
\end{notice}
```

again the first rendered admonition is buggy -- the title is always missing

It seems that there is another issue.

**Comment by rexut** [comment]
To reproduce the wrong LaTeX output from https://github.com/sphinx-doc/sphinx/issues/3117#issuecomment-258613364 I'll provide this test repo:

https://github.com/rexut/sphinx-admonition-test

Now I'll investigate the issue with some older and latest Sphinx releases.

**Comment by jfbu** [comment]
Yes thank you for report I confirm it, but the `index.rst` (I copied your rst exactly) is missing sectioning (and toctree directive). Adding 

```
A section
---------
```

on top of file removes the problem. 

@tk0miya  what do you think ? is it legal to have rst source with no sectioning ?

**Comment by jfbu** [comment]
actually  I copied-pasted from above (I had not seen your message about github repo) hence I don't even have the inline literal. We have two exactly identical generic admonitions. The latex output is wrong for the first one. There is no problem if the rst source contains a section title.

**Comment by jfbu** [comment]
pseudoxml, html, texinfo, epub have no issue. Hence perhaps indeed there is a latex writer or builder problem. But I don't know enough the internals to judge.

**Comment by rexut** [comment]
> But I don't know enough the internals to judge.

Same to me :-)

**Comment by tk0miya** [comment]
It seems latex writer handles first title node.
https://github.com/sphinx-doc/sphinx/blob/1.4.8/sphinx/writers/latex.py#L768

I don't know what is expected here. But from admonition side, this is not unexpected.
So I marked this as a bug.

> what do you think ? is it legal to have rst source with no sectioning ?

Sorry, I don't know that.
But latex writer skips the first section title. So the writer gives it special treatment.
In other words, it is implicit expectation.

**Comment by rexut** [comment]
> It seems latex writer handles first title node.
> https://github.com/sphinx-doc/sphinx/blob/1.4.8/sphinx/writers/latex.py#L768

This is also my observations. I would prefer to rearrange the priority in the if-elif-sequence. Before play with this idea, I've create (and pulled) a unit test that raise this issue.

For clarification of this issue: The origin of this discussion was a problem with the [sphinxcontrib-traceables](https://github.com/t4ngo/sphinxcontrib-traceables) extension. That will be fixed there by a separate PR. Here I was stumbled over a hidden really rare fault condition with the admonition directive that lives outside of any section context.

**Comment by tk0miya** [comment]
In my short investigation:
The condition `elif self.this_is_the_title:` appeared first is in 3d74adea; the first version of LaTeX writer.
At this era, LaTeX writer had only handled title nodes under sections.

BTW, the first node is used for `\title{...}` macro of the document.
I think title node under section is befitting to the usage.

So I decided to move the condition into `elif isinstance(node.parent, nodes.section)` block.

**Comment by tk0miya** [comment]
Fixed at 	43197de.
Thank you for reporting!
